src/client_side.cc

```diff
@@ -4117,23 +4121,25 @@ ConnStateData::checkLogging()
 }
 
 bool
-ConnStateData::mayTunnelUnsupportedProto()
+ConnStateData::preserveDataForTunnellingUnsupportedProto()
 {
-    if (!Config.accessList.on_unsupported_protocol)
-        return false;
-
 #if USE_OPENSSL
     // Tunneling may be possible when parsing the TLS client handshake
     // and when parsing the first HTTP request on a bumped connection.
-    if (sslBumpMode != Ssl::bumpEnd && parsedBumpedRequestCount <= 1)
-        return true;
+    bool firstBumpedRequest = sslBumpMode != Ssl::bumpEnd && parsedBumpedRequestCount <= 1;
```

Comment:
The comment says `Tunneling may be possible when A and when B`. The boolean variable is called `whenB`. What happened to A?! I do not know what the exact solution is, but the code (including comments and variable names) does not make sense right now. Something needs to be changed!

> There is not a safe check for "while parsing the TLS handshake".

I do not understand what that means or implies in this specific context. The comment is lying about us checking that condition? The code is unsafe? The comment is irrelevant to the code it is covering? The existing check covers both cases and cannot (or should not) be split into two (but should be renamed)? The correct solution to address this problem may depend on the correct answers to those questions. I do not know those answers, but you probably do.

> While we are parsing tls handshake we are always preserve data to be able to splice if required by configuration.

OK.

>  In all cases the mayTunnelUnsupportedProto is called after the sslBumpMode is set and while parsedBumpedRequestCount is less or equal to 1.

I do not know what you mean by "all cases". Since we are checking that condition, that condition is probably not true in some cases. The variable name should reflect what the condition means.

Reply:
> The comment says Tunneling may be possible when A and when B. The boolean variable is called whenB. What happened to A?

The A (was/)is handled setting directly inside the ConnStateData::parseTlsHandshake() method.

> Or that this method should be called preserveHttpBytesForTunnellingUnsupportedProto because it only deals with plain HTTP and decrypted HTTPS bytes?

Yes. I used this name in my commit e757c76